hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/avro/HudiAvroParquetReader.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.ArrayWritable;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hudi.hadoop.HoodieColumnProjectionUtils;
+import org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils;
+import org.apache.parquet.avro.AvroReadSupport;
+import org.apache.parquet.avro.AvroSchemaConverter;
+import org.apache.parquet.format.converter.ParquetMetadataConverter;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.ParquetInputSplit;
+import org.apache.parquet.hadoop.ParquetRecordReader;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.schema.MessageType;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.parquet.hadoop.ParquetInputFormat.getFilter;
+
+public class HudiAvroParquetReader extends RecordReader<Void, ArrayWritable> {
+
+  private final ParquetRecordReader<GenericData.Record> parquetRecordReader;
+
+  public HudiAvroParquetReader(InputSplit inputSplit, Configuration conf) throws IOException {
+    AvroReadSupport avroReadSupport = new AvroReadSupport<>();
+    // if exists read columns, we need to filter columns.
+    List<String> readColNames = Arrays.asList(HoodieColumnProjectionUtils.getReadColumnNames(conf));
+    if (!readColNames.isEmpty()) {
+      // get base schema
+      ParquetMetadata fileFooter =
+              ParquetFileReader.readFooter(conf, ((ParquetInputSplit) inputSplit).getPath(), ParquetMetadataConverter.NO_FILTER);
+      MessageType messageType = fileFooter.getFileMetaData().getSchema();
+      Schema baseSchema = new AvroSchemaConverter(conf).convert(messageType);
+      // filter schema for reading
+      final Schema filterSchema = Schema.createRecord(baseSchema.getName(),
+              baseSchema.getDoc(), baseSchema.getNamespace(), baseSchema.isError(),
+              baseSchema.getFields().stream()
+                      .filter(f -> readColNames.contains(f.name()))
+                      .map(f -> new Schema.Field(f.name(), f.schema(), f.doc(), f.defaultVal()))
+              .collect(Collectors.toList()));
+      avroReadSupport.setAvroReadSchema(conf, filterSchema);
+      avroReadSupport.setRequestedProjection(conf, filterSchema);
+    }
+    parquetRecordReader = new ParquetRecordReader<>(avroReadSupport, getFilter(conf));
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+    parquetRecordReader.initialize(split, context);
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    return parquetRecordReader.nextKeyValue();
+  }
+
+  @Override
+  public Void getCurrentKey() throws IOException, InterruptedException {
```

Comment:
Void？

Reply:
@xiarixiaoyao 

I mainly refer to `org.apache.hadoop.hive.ql.io.parquet.read.ParquetRecordReaderWrapper` in `hive-exec` package.
https://github.com/apache/hive/blob/branch-2.3/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/read/ParquetRecordReaderWrapper.java#L47
```java
  private org.apache.hadoop.mapreduce.RecordReader<Void, ArrayWritable> realReader;
```

It should be declared that paruqet reader actually only uses value to process row data.

Reply:
got it